code/datums/components/overlay_lighting.dm

```diff
@@ -0,0 +1,306 @@
+		COMSIG_ATOM_SET_LIGHT_ON,
+		COMSIG_ATOM_SET_LIGHT_FLAGS,
+		))
+	if(overlay_lighting_flags & LIGHTING_ON)
+		turn_off()
+	return ..()
+
+
+/datum/component/overlay_lighting/Destroy()
+	set_holder(null)
+	clean_old_turfs()
+	visible_mask = null
+	return ..()
+
+
+///Clears the affected_turfs lazylist, removing from its contents the effects of being near the light.
+/datum/component/overlay_lighting/proc/clean_old_turfs()
+	for(var/turf/lit_turf as anything in affected_turfs)
+		lit_turf.dynamic_lumcount -= lum_power
+	affected_turfs = null
+
+
+///Populates the affected_turfs lazylist, adding to its contents the effects of being near the light.
+/datum/component/overlay_lighting/proc/get_new_turfs()
+	if(!current_holder)
+		return
+	LAZYINITLIST(affected_turfs)
+	if(range <= 2)
+		//Range here is 1 because actual range of lighting mask is 1 tile even if it says that range is 2
+		for(var/turf/lit_turf in RANGE_TURFS(1, current_holder.loc))
+			lit_turf.dynamic_lumcount += lum_power
+			affected_turfs += lit_turf
+	else
+		for(var/turf/lit_turf in view(lumcount_range, get_turf(current_holder)))
+			lit_turf.dynamic_lumcount += lum_power
+			affected_turfs += lit_turf
+
+
+///Clears the old affected turfs and populates the new ones.
+/datum/component/overlay_lighting/proc/make_luminosity_update()
+	clean_old_turfs()
+	if(!isturf(current_holder?.loc))
+		return
+	get_new_turfs()
+
+
+///Adds the luminosity and source for the afected movable atoms to keep track of their visibility.
+/datum/component/overlay_lighting/proc/add_dynamic_lumi(atom/movable/affected_movable)
+	LAZYSET(affected_movable.affected_movable_lights, src, lumcount_range + 1)
+	affected_movable.underlays += visible_mask
+	affected_movable.update_dynamic_luminosity()
+
+
+///Removes the luminosity and source for the afected movable atoms to keep track of their visibility.
+/datum/component/overlay_lighting/proc/remove_dynamic_lumi(atom/movable/affected_movable)
+	LAZYREMOVE(affected_movable.affected_movable_lights, src)
+	affected_movable.underlays -= visible_mask
+	affected_movable.update_dynamic_luminosity()
+
+
+///Called to change the value of current_holder.
+/datum/component/overlay_lighting/proc/set_holder(atom/movable/new_holder)
+	if(new_holder == current_holder)
+		return
+	if(current_holder)
+		if(overlay_lighting_flags & LIGHTING_ON)
+			remove_dynamic_lumi(current_holder)
+	current_holder = new_holder
+	if(new_holder == null)
+		clean_old_turfs()
+		return
+	if(overlay_lighting_flags & LIGHTING_ON)
+		add_dynamic_lumi(new_holder)
+
+
+///Used to determine the new valid current_holder from the parent's loc.
+/datum/component/overlay_lighting/proc/check_holder()
+	var/atom/movable/movable_parent = parent
+	if(isturf(movable_parent.loc))
+		set_holder(movable_parent)
+		return
+	var/atom/inside = movable_parent.loc //Parent's loc
+	if(isnull(inside))
+		set_holder(null)
+		return
+	if(isturf(inside.loc))
+		set_holder(inside)
+		return
+	set_holder(null)
```

Comment:
For the issue where lights pre-equipped in cryopods don't work, now the light will still attach even if we are deeper inside something's contents (cryopod).
```suggestion
	if(!movable_parent || !get_turf(movable_parent))
		set_holder(null)
		return
	// Try only going up to parent one level
	if(!isturf(movable_parent.loc) && !ismob(movable_parent))
		movable_parent = movable_parent.loc
	set_holder(movable_parent)
```
You may also consider adding more checks to either movable_parent or movable_parent.loc for any other scenarios where you don't want parent getting illuminated when inserted. E.g. `&& !istype(movable_parent.loc, /obj/structure/closet)`